### PR TITLE
replace strong-agent with appmetrics

### PR DESCRIPTION
**Pull request by sam-github**
connected to https://github.com/strongloop-internal/scrum-nodeops/issues/1465

### CODE REVIEW

File: lib/adapter.js [hunk 1]
```diff
@@ -0,0 +1,494 @@
+/*******************************************************************************
+ * Copyright 2014, 2015, 2016 IBM Corp.
+ * 
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ * 
+ * http://www.apache.org/licenses/LICENSE-2.0
```
Comment (sam-github) [inline lib/adapter.js]: everything else in this repo is MIT, do we need to change it to apache-2.0, or can we change this to MIT? I think having multiple licenses should be avoided if possible.

**Comment by sam-github** [comment]
@slnode test please

File: lib/targetctl.js [hunk 2]
```diff
@@ -210,15 +221,17 @@ function onRequest(req, callback) {
...
     }
   } catch (err) {
     rsp.error = err.message;
+    callback(rsp);
+
   }
 
   debug('request %s => response %s', debug.json(req), debug.json(rsp));
 
-  callback(rsp);
```
Comment (sam-github) [inline lib/targetctl.js]: why was this and the code that used it changed? It doesn't seem related or necessary.

File: lib/watcher/cpu-profile.js [hunk 3]
```diff
@@ -28,7 +28,7 @@ exports.worker = function(handle) {
       return;
 
     try {
-      var profile = agent().metrics.stopCpuProfiling();
+      agent().metrics.stopCpuProfiling(function(profile) {
```
Comment (sam-github) [inline lib/watcher/cpu-profile.js]: Please indent the function body

File: package.json [hunk 4]
```diff
@@ -51,7 +51,9 @@
     "strong-statsd": "^2.0.0",
     "strong-trace": "^1.2.0",
     "strong-url-defaults": "^1.2.0",
-    "strongloop-license": "^1.3.0"
+    "strongloop-license": "^1.3.0",
+    "appmetrics":"RuntimeTools/appmetrics#adapter"
```
Comment (sam-github) [inline package.json]: sort the dependencies (npm will do it for you if you do `npm install --save RuntimeTools/appmetrics#adapter` - with a recent enough npm.

File: lib/adapter.js [hunk 5]
```diff
@@ -0,0 +1,494 @@
...
+var path = require('path');
+
+appmetrics.enable('eventloop');
+var amapi = appmetrics.monitor();
+
+// Global objects to store metric data
+var eventLoopMetrics = {count: 0, averageRunningTotal: 0};
+var cpuMetrics = {system: 0, user: 0, count: 0};
+var httpMetrics = {count: 0};
+var probeList = ["leveldown", "memcached", "mongo", "mysql", "oracle", "oracledb", "postgres", "redis", "socketio", "strong-oracle"];
```
Comment (sam-github) [inline lib/adapter.js]: Run `npm run pretest` to lint the code (this line is too long).

File: lib/adapter.js [hunk 6]
```diff
@@ -0,0 +1,494 @@
...
+// ---------------------------------------------
+// Public functions to be exported
+// ---------------------------------------------
+
+// Metrics API Use function
+function use(callback) {
+    // Set up listeners for all the metrics here
+    initialise();
+
+    // The callback needs to fire every 60 seconds with data from the last 60
```
Comment (sam-github) [inline lib/adapter.js]: you should probably remove mentions of explicit intervals in the comments, they tend to drift out of sync with the code. The actual interval looks to be 10 seconds.

File: lib/adapter.js [hunk 7]
```diff
@@ -0,0 +1,494 @@
...
+var gcMetrics = {used: 0, count: 0};
+var messageMetrics = {};
+
+// Keep track of Cpu Profiling
+var profiling = false;
+var objectTrackingEnabled = false;
+var lastObjectHistogram = {};
+
+// Callback interval
+var callbackInterval = 10000;
```
Comment (sam-github) [inline lib/adapter.js]: is it intended that these intervals be different, or is this debug code?

Its useful to have them be configurable for test purposess, I'd define these both to be `process.env.STRONGLOOP_BASE_INTERVAL || 15000`

File: lib/adapter.js [hunk 8]
```diff
@@ -0,0 +1,494 @@
...
+    initialise();
+
+    // The callback needs to fire every 60 seconds with data from the last 60
+	// seconds
+    setInterval(function() {
+        // Get data from last 60 seconds
+        var eventLoopData = getEventLoopData();
+        var cpuData = getCpuData();
+        var httpData = getHTTPData();
+		var probeData;
```
Comment (sam-github) [inline lib/adapter.js]: you are using hardware tabs, looks like, with a non-standard tab width. use spaces (eslint will tell you about this when you run the pretest)

File: lib/adapter.js [hunk 9]
```diff
@@ -0,0 +1,494 @@
...
+var objectTrackingEnabled = false;
+var lastObjectHistogram = {};
+
+// Callback interval
+var callbackInterval = 10000;
+var objectTrackingCallbackInterval = 15000;
+
+// ---------------------------------------------
+// Public functions to be exported
+// ---------------------------------------------
```
Comment (sam-github) [inline lib/adapter.js]: nit: please avoid this bar-style comments, the text comment is sufficient

File: lib/adapter.js (hunk 8 above)
Comment (sam-github) [inline lib/adapter.js]: please define your vars closer to where they are used, it makes it easier to read the code without paging back and forth

File: lib/adapter.js [hunk 10]
```diff
@@ -0,0 +1,494 @@
...
+				probeData = getProbeData(probeList[i]);
+				callback(probeList[i] + ".count", probeData.count);
+				callback(probeList[i] + ".average", probeData.average);
+				callback(probeList[i] + ".minimum", probeData.minimum);
+				callback(probeList[i] + ".maximum", probeData.maximum);
+			}
+		}
+
+		//if (usedHeap > 0) {
+			callback("gc.heap.used", usedHeap);
```
Comment (sam-github) [inline lib/adapter.js]: I'll have to check, but IIRC, we only emitted this if there had been a gc event. But my memory is fabulously bad.

File: lib/adapter.js [hunk 11]
```diff
@@ -0,0 +1,494 @@
...
+		callback("heap.total", heapData.total);
+		callback("heap.used", heapData.used);
+
+        if (messageData != null) {
+            callback("messages.out.count", messageData.sent);
+            callback("messages.in.count", messageData.received);  
+        }
+		
+		
+	}, callbackInterval);
```
Comment (sam-github) [inline lib/adapter.js]: you'll have to `unref()` your timers or node won't exit, you'll probably run into this when you run the tests.

File: lib/adapter.js [hunk 12]
```diff
@@ -0,0 +1,494 @@
...
+				}
+
+			}, objectTrackingCallbackInterval);
+}
+
+function start() {
+
+}
+
+function on() {
```
Comment (sam-github) [inline lib/adapter.js]: don't you need this?

File: lib/adapter.js [hunk 13]
```diff
@@ -0,0 +1,494 @@
...
+	
+}
+
+function internal_on() {
+	
+}
+
+
+function profile(arg1, arg2, arg3) {
```
Comment (sam-github) [inline lib/adapter.js]: don't you need this? its what starts profiling, if its not called, the profiler should not be running. appmetrics has a similar function, if it is required but not started it doesn't do anything

File: lib/adapter.js [hunk 14]
```diff
@@ -0,0 +1,494 @@
...
+		appmetrics.enable('profiling');
+
+	}
+}
+
+function stopCpuProfiling(callback) {
+	if (profiling) {
+		
+		// Listen for JSON coming back from appmetrics
+		amapi.on('profiling', function(data) {
```
Comment (sam-github) [inline lib/adapter.js]: `once()` - or the callbacks will accumulate, and every subsequent stop will attempt to call every stop callback ever registered

File: lib/adapter.js [hunk 15]
```diff
@@ -0,0 +1,494 @@
...
+			callback(JSON.stringify(data));
+		});
+
+		profiling = false;
+		appmetrics.disable('profiling');
```
Comment (sam-github) [inline lib/adapter.js]: `disable('profiling')` causes the `profiling` event?

File: lib/adapter.js [hunk 16]
```diff
@@ -0,0 +1,494 @@
...
+	}
+}
+
+// Object tracking
+function startTrackingObjects() {
+	objectTrackingEnabled = true;
```
Comment (sam-github) [inline lib/adapter.js]: I'll keep reading, but it seems this would imply that object tracking is always on in appmetrics, won't that be pretty expensive?

File: lib/adapter.js [hunk 17]
```diff
@@ -0,0 +1,494 @@
...
+	amapi.on('cpu', function(cpu) {
+		cpuMetrics.system += (cpu.system * 100);
+		cpuMetrics.user += (cpu.process * 100);
+		cpuMetrics.count++;
+	});
+}
+
+function createHeapListener() {
+	amapi.on('heap', function(res) {
+		heapMetrics.total = +res.total;
```
Comment (sam-github) [inline lib/adapter.js]: was this intended to be `+=`?

File: lib/adapter.js [hunk 18]
```diff
@@ -0,0 +1,494 @@
...
+		heapMetrics.used = +res.used;
+		heapMetrics.count++;
+	});
+}
+
+function createGCListener() {
+	amapi.on('gc', function(res) {
+		gcMetrics.used = +res.used;
```
Comment (sam-github) [inline lib/adapter.js]: this, too, was this supposed to be `+=`, or is it a conversion from string to number?

File: lib/adapter.js [hunk 19]
```diff
@@ -0,0 +1,494 @@
...
+	amapi.on('http', function(res) {
+		// All we need to do is increment the count
+		httpMetrics.count++;
+	});
+}
+
+function createMessageListener() {
+	// Map the event types from axon and strong mq to whether they 
+	// were send or receive events.
+	var incrementCounts = function (data) {
```
Comment (sam-github) [inline lib/adapter.js]: please just define it as a function: `function incrementCounts(data) {`, and define it after use at the end of the function - its more understandable to read code in the order it is executed: `.on(blah, fu); function fu() { ...} `

File: lib/adapter.js [hunk 20]
```diff
@@ -0,0 +1,494 @@
...
+            }
+		}
+	}
+	amapi.on('strong-mq', incrementCounts);
+	amapi.on('axon', incrementCounts);
+}
+
+function createObjectListener() {
+	amapi.on('objectType', function(objectType) {
+
```
Comment (sam-github) [inline lib/adapter.js]: I guess the object stuff is incomplete?

File: lib/adapter.js [hunk 21]
```diff
@@ -0,0 +1,494 @@
...
+	});
+	amapi.on('objectSize', function (objectSize) {
+		
+	});
+}
+
+function createProbeListener(probeName){
+	amapi.on(probeName, function(res){
+		var duration = res.duration;
+		var currentMetrics;
```
Comment (sam-github) [inline lib/adapter.js]: unused, incomplete or left over?

File: lib/adapter.js [hunk 22]
```diff
@@ -0,0 +1,494 @@
...
+		
+		// If no metrics object exists, create one
+		if (probeMetrics[probeName] == undefined){
```
Comment (sam-github) [inline lib/adapter.js]: `if (!probeMetrics[probeName]) {`

File: lib/adapter.js [hunk 23]
```diff
@@ -0,0 +1,494 @@
...
+
+	// Reset Count
+	httpMetrics.count = 0;
+
+    return httpEvent
+}
+
+function getProbeData(probeName){
+	
+	//Get data for given probe
```
Comment (sam-github) [inline lib/adapter.js]: `// Get`

File: lib/adapter.js [hunk 24]
```diff
@@ -0,0 +1,494 @@
...
+	var probeData = probeMetrics[probeName];
+	var count = probeData.count;
+	var average;
```
Comment (sam-github) [inline lib/adapter.js]: `var average = count ? probeData.runningTotal / count : 0` would compress 5 lines into 1 (we aren't members of the "ternary operators are evil" cult).

File: lib/adapter.js [hunk 25]
```diff
@@ -0,0 +1,494 @@
...
+		gcMetrics.used = 0;
+		gcMetrics.count = 0;
+	} else {
+		usedHeap = -1;
+	}
+	return usedHeap;
+}
+
+function getMessageData() {
+    if (messageMetrics.sent != undefined && messageMetrics.received != undefined) {
```
Comment (sam-github) [inline lib/adapter.js]: Won't this mean that an app that only sends, or only receives, won't ever get metrics reported?

File: lib/adapter.js [hunk 26]
```diff
@@ -0,0 +1,494 @@
...
+        var oldMessageMetrics = messageMetrics;
+        messageMetrics = {};
+        return oldMessageMetrics;
+    }
+    return null;
```
Comment (sam-github) [inline lib/adapter.js]: redundant line, functions return `undefined` already (unless you need to distinguish between null and undefined, but I don't think you do)

File: lib/adapter.js [hunk 27]
```diff
@@ -0,0 +1,494 @@
...
+	appName : path.basename(process.argv[1]),
+	key : 'testkey'
+};
+
+var internal = {
+		'on' : internal_on
+};
+
+
+module.exports.use = use;
```
Comment (sam-github) [inline lib/adapter.js]: `exports.use = use;` (`module.` is redundant), and place at top of file, underneath the requires, so its easy to see at a glance what a module exports.

File: lib/adapter.js [hunk 28]
```diff
@@ -0,0 +1,494 @@
...
+module.exports.metrics = metrics;
+module.exports.config = config;
+module.exports.profile = profile;
+module.exports.start = start;
+module.exports.configure = configure;
+module.exports.on = on;
+module.exports.internal = internal;
+
```
Comment (sam-github) [inline lib/adapter.js]: eslint will probably warn you about this, but delete trailing blank lines, please.

**Comment by sam-github** [comment]
^--- tests don't have a hope of passing until `npm run pretest` passes. I just looked over the code, I'll find time later to run it, but I believe that `slc arc` is functioning, except the object tracking seems not present, or did I read to quickly?

File: lib/adapter.js (hunk 6 above)
Comment (DanCunnington) [inline lib/adapter.js]: Hi Sam, ok - we will update the comment. The interval was moved to 10 seconds for testing but we will put that back to 60. Thanks

File: lib/adapter.js (hunk 1 above)
Comment (tobespc) [inline lib/adapter.js]: we will look into this and check with project management as to what license we should be using

File: lib/targetctl.js (hunk 2 above)
Comment (tobespc) [inline lib/targetctl.js]: we had to change this due to the way appmetrics returned the cpu profiling as a callback.  The case 'stop-cpu-profiling' needs to wait for the callback before returning and so we had to move the callback(rsp) that was at the end of the code to be triggered in each case statement.  Otherwise, the 'stop-cpu-profiling' would return without waiting for the callback and an error would occur due to an empty profile

File: lib/adapter.js (hunk 7 above)
Comment (tobespc) [inline lib/adapter.js]: These variables do need to be different values, 60 seconds for the metrics and 15 seconds for the object tracking from the strong-agent api.  Is this the case ?

File: lib/adapter.js (hunk 10 above)
Comment (tobespc) [inline lib/adapter.js]: will fix this

File: lib/adapter.js (hunk 12 above)
Comment (tobespc) [inline lib/adapter.js]: appmetrics starts automatically, we don't need to say start,  This is a blank stub to make it work with strong-supervisor

File: lib/adapter.js (hunk 13 above)
Comment (tobespc) [inline lib/adapter.js]: the function "startCpuProfiling" below this starts profiling.  All these blank stubs are there for compatibility with Strong-supervisor making calls to these functions.

File: lib/adapter.js (hunk 15 above)
Comment (tobespc) [inline lib/adapter.js]: yes, 'disable' causes appmetrics to stop profiling and return all the data.  Its our wording

File: lib/adapter.js (hunk 16 above)
Comment (tobespc) [inline lib/adapter.js]: objecttracking will start when 'startTrackingObject' is called.  The process loop in the "function use(callback)" will collect the data whilst the boolean is set and once 'stopTrackingObject' is called, we will stop collecting it

File: lib/adapter.js (hunk 21 above)
Comment (tobespc) [inline lib/adapter.js]: this is called for each probe in our list

File: lib/adapter.js (hunk 24 above)
Comment (tobespc) [inline lib/adapter.js]: we are and our membership is due :-)   (we'll make the changes)

File: lib/adapter.js (hunk 25 above)
Comment (tobespc) [inline lib/adapter.js]: correct, changing....

File: lib/adapter.js (hunk 13 above)
Comment (sam-github) [inline lib/adapter.js]: sorry, terminology crash, profiling != cpu profiling :-(

Anyhow, supervisor needs a way to NOT run appmetrics when  called with --no-profile, in which case this function, `profile()`, is not called, and no profiling occurs.

besides avoiding appmetrics overhead, this is pretty much a mandatory feature for trying to figure out if a bug or behavioural oddity we see was caused by the metrics/instrumentation (turn it off, see how it still works).

IIRC, the act of requiring appmetrics causes it to start, no choice?

File: lib/adapter.js (hunk 21 above)
Comment (sam-github) [inline lib/adapter.js]: The line I commented on, `var currentMetrics` is an unused variable.

File: lib/adapter.js (hunk 13 above)
Comment (tunniclm) [inline lib/adapter.js]: Quick note: there is a bug in appmetrics that means it starts on `require()`, as documented in the readme, it _should_ only start when either `start()` or `monitor()` is called. I actually don't think we have an issue raised for this!

Comment (seabaylea) [inline lib/adapter.js]: Yes - that was/is the intended design - with `monitor()` implicitly calling `start()` if monitoring isn't already started.

**Comment by sam-github** [comment]
@tobespc Fixing some of linting had the effect of hiding some of the comments:

- [ ] `npm run pretest` shows lint errors, please fix
- [ ] appmetrics is running even in `--no-profile` mode
- [ ] exports are on bottom of file, they should be on top

File: lib/adapter.js [hunk 29]
```diff
@@ -0,0 +1,469 @@
...
+    }
+
+    if (usedHeap > 0) {
+      callback('gc.heap.used', usedHeap);
+    }
+
+    callback('heap.total', heapData.total);
+    callback('heap.used', heapData.used);
+
+    if (messageData !== null) {
```
Comment (sam-github) [inline lib/adapter.js]: `messageData` is never `null` after a recent refactor, it is always `{}` at least, perhaps with some properties, so I think the message in/out counts will be reported as `undefined` often.

File: lib/watcher/cpu-profile.js [hunk 30]
```diff
@@ -28,14 +28,14 @@ exports.worker = function(handle) {
...
     try {
-      var profile = agent().metrics.stopCpuProfiling();
-
-      handle.emit({
-        cmd: 'cpu:profile-data',
-        stalls: stalls,
-        stallout: stallout,
-        profile: profile,
-        // stopTime: Date.now(), XXX could be useful, but doesn't have a user
+      agent().metrics.stopCpuProfiling(function(profile) {
```
Comment (sam-github) [inline lib/watcher/cpu-profile.js]: This code is still wrapped in try/catch. but I believe that it does not throw, so the try/catch should be removed.

The callback structure is unusual, it asynchronous (I assume, because strong-agent uses a direct return, and you needed to change it to a callback). However, the callback is not err-first. This means that this call *cannot* fail? What about when no profile is active? Or... will it throw when no profile is active?

Both throwing exceptions *and* calling back is possible in Node.js, but its atypical, used mostly for severe errors in the construction of the code (args missing or wrong time), not usually for run-time errors.

Net: I think that the callback should be err-first.

Comment (sam-github) [inline lib/watcher/cpu-profile.js]: http://fredkschott.com/post/2014/03/understanding-error-first-callbacks-in-node-js/

File: lib/adapter.js [hunk 31]
```diff
@@ -0,0 +1,497 @@
...
+exports.metrics = metrics;
+exports.config = config;
+exports.profile = profile;
+exports.start = start;
+exports.configure = configure;
+exports.on = on;
+exports.internal = internal;
+
+var appmetrics; // = require('appmetrics') if started
+var appmetricsStarted = false; // XXX(sam) === !!appmetrics, so redundant?
```
Comment (sam-github) [inline lib/adapter.js]: @tobespc

File: lib/adapter.js [hunk 32]
```diff
@@ -0,0 +1,497 @@
...
+var amapi; // = appmetrics.monitor() XXX(sam) so maybe call this 'monitor'?
```
Comment (sam-github) [inline lib/adapter.js]: @tobespc

**Comment by sam-github** [comment]
@tobespc I rebased and squashed

File: lib/adapter.js [hunk 33]
```diff
@@ -0,0 +1,497 @@
...
+
+// Keep track of Cpu Profiling
+var profiling = false;
+var objectTrackingEnabled = false;
+var lastObjectHistogram = {};
+
+// Callback interval
+var callbackInterval = process.env.STRONGLOOP_BASE_INTERVAL || 60000;
+var objectTrackingCallbackInterval =
+  process.env.STRONGLOOP_OBJECT_TRACKING_INTERVAL || 15000;
```
Comment (sam-github) [inline lib/adapter.js]: I'm looking closer, these duplicate/conflict with bin/sl-run.js:agentOptions.interval

File: lib/adapter.js [hunk 34]
```diff
@@ -0,0 +1,497 @@
...
+  setInterval(function() {
+    if (objectTrackingEnabled) {
+      var histogram = appmetrics.getObjectHistogram();
+      var objectCounts = compareHistograms(lastObjectHistogram, histogram);
+      lastObjectHistogram = histogram;
+      // Get object stats
+      for (var type in objectCounts) {
+        var tuple = objectCounts[type];
+        // Don't output events for no change?
+        // XXX(toby) - What's the correct behaviour here?
```
Comment (sam-github) [inline lib/adapter.js]: @tobespc , I'm looking into this

File: lib/adapter.js [hunk 35]
```diff
@@ -0,0 +1,497 @@
...
+
+  // Reset values
+  cpuMetrics.system = 0;
+  cpuMetrics.user = 0;
+  cpuMetrics.count = 0;
+
+  return cpuEvent;
+}
+
+function getHTTPData() {
```
Comment (sam-github) [inline lib/adapter.js]: camelCase

File: lib/adapter.js [hunk 36]
```diff
@@ -0,0 +1,497 @@
...
+var metrics = {
+  startCpuProfiling: startCpuProfiling,
+  stopCpuProfiling: stopCpuProfiling,
+  startTrackingObjects: startTrackingObjects,
+  stopTrackingObjects: stopTrackingObjects
+};
+
+var config = {
+  hostname: os.hostname(),
+  appName: path.basename(process.argv[1]),
```
Comment (sam-github) [inline lib/adapter.js]: (for sam) look into appName and key, what they are (appName here will be sl-run.js, but that is not the name of the application being run)

File: lib/adapter.js [hunk 37]
```diff
@@ -0,0 +1,501 @@
...
+    appmetrics = require('appmetrics');
+    appmetrics.enable('eventloop');
+    amapi = appmetrics.monitor();
+    initialise();
+    appmetricsStarted = true;
+  }
+}
+
+function on() {
+  // XXX(sam) Unimplemented features:
```
Comment (sam-github) [inline lib/adapter.js]: its not clear to me whether we need full support for all strong-agent features, but if we do, these are missing

File: lib/capabilities.js [hunk 38]
```diff
@@ -41,7 +41,7 @@ var capabilities = {
       },
       {
         check: function(cb) {
-          if (agent().internal.supports.watchdog) {
+          if (agent().internal.supports.watchdog) { // XXX(sam) appmetrics?
```
Comment (sam-github) [inline lib/capabilities.js]: appmetrics won't run without its compiled addon, so some of the checks in here can be hard-coded to true, and some other capabilities (tracing) aren't supported by appmetrics now, so will be false. `agent().internal.supports.watchdog` will probably fail, there is no `.supports` in the adapter shim (but the conditional should just be rewritten)

**Comment by sam-github** [comment]
Can someone update the appmetrics on this branch? I believe it is segraulting:

```
sn/strong-pm (master $ u=) % sl-pm
sl-pm(15221): StrongLoop PM v5.2.1 (API v6.1.0) on port `8701`
sl-pm(15221): Base folder `/home/sam/.strong-pm`
sl-pm(15221): Applications on port `3000 + service ID`
Browse your REST API at http://127.0.0.1:8701/explorer
Run request for commit "default/npm-pack" on current (none)
Start Runner: commit default/npm-pack
*** Error in `/home/sam/.nvm/versions/node/v4.2.1/bin/node': free(): invalid pointer: 0x0000000003235b78 ***
Unexpected exit with SIGABRT from Runner: commit default/npm-pack
Start Runner: commit default/npm-pack
*** Error in `/home/sam/.nvm/versions/node/v4.2.1/bin/node': free(): invalid pointer: 0x000000000351b0d8 ***
Unexpected exit with SIGABRT from Runner: commit default/npm-pack
... repeated forever
```

That's as far as I got in my attempt to run with arc to confirm backwards compat.

Also, the tests are failing, looks like because of the capabilities things I commented on, but also the invalid pointer, possibly other things.

Please run `npm test`.

**Comment by tobespc** [comment]
regarding appmetrics being an older version, we are going to update the master branch to include the adapter changes this week.  We will then modify the dependancy to pull from master rather than the adaptor branch

**Comment by tobespc** [comment]
appmetrics adaptor branch has now been updated with the latest code from master

**Comment by sam-github** [comment]
@tobespc Sorry about the exports snafu, should be fixed now, and I added a regression test.

Current state:

- `npm test` still fails
- lib/capabilities.js not ported to use appmetrics
- node still segfaulting after update (I've confirmed its pulled the latest from your adapter branch):

```
$ tap test/test-*.js
...
test/test-ipcctl-notifications.js supervisor (16378) logging to '_test-16373-run.log'
test/test-ipcctl-notifications.js ........*** Error in `/home/sam/.nvm/versions/node/v4.2.1/bin/node': free(): invalid pointer: 0x0000000003181958 ***
^C
sn/strong-supervisor (appmetrics-adapter u=) % uname -a
Linux samtu 4.2.0-35-generic #40-Ubuntu SMP Tue Mar 15 22:15:45 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by sam-github** [comment]
Looks like this is a known bug: https://github.com/RuntimeTools/appmetrics/issues/190

File: lib/adapter.js [hunk 39]
```diff
@@ -203,6 +205,13 @@ function profile() {
...
 // configure(agentOptions)
 function configure() {
+  // FIXME - does appmetrics maintain a notion of the application's name?  If it
+  // does not, then supervisor can decide. strong-agent did, and supervisor
+  // didn't want to duplicate its logic, or risk its idea of the app's name
+  // being different from strong-agents.
+  return {
+    appName: 'my-app',
+  };
 }
```
Comment (tobespc) [inline lib/adapter.js]: appmetrics does not maintain a notion of the applications name

**Comment by sam-github** [comment]
@slnode test please

**Comment by sam-github** [comment]
@rmg do we need toby to sign the CLA?

**Comment by rmg** [comment]
@sam-github shouldn't.. but @tobespc might need to sign in to https://cla.strongloop.com/ so that it can recognize he's an org member.

**Comment by sam-github** [comment]
@rmg PTAL, I would like to merge

File: lib/adapter.js [hunk 40]
```diff
@@ -0,0 +1,548 @@
...
+  received: 0,
+};
+
+// Keep track of Cpu Profiling
+var profiling = false;
+var objectTrackingEnabled = false;
+var lastObjectHistogram = {};
+
+// Callback interval
+var callbackInterval = process.env.STRONGLOOP_BASE_INTERVAL || 60000;
```
Comment (kjdelisle) [inline lib/adapter.js]: If you're calling the other one `objectTrackingCallbackInterval`, shouldn't this one be `baseCallbackInterval`?

File: lib/tracer.js [hunk 41]
```diff
@@ -35,9 +29,12 @@ function tracer() {
 function start() {
   assert(!TRACER);
   // This is delayed to allow process.env to be modified up until start time.
-  var config = configure(null, null, {}, process.env);
-  options.accountKey = config.appName;
-  if (options.accountKey)
+  options.accountKey = app.name();
+  if (options.accountKey) {
+    // This is delayed becausing requiring appmetrics causes it to monkey-patch,
```
Comment (kjdelisle) [inline lib/tracer.js]: `becausing` -> `because`

File: lib/targetctl.js [hunk 42]
```diff
@@ -166,14 +166,17 @@ function onRequest(req, callback) {
...
       case 'stop-cpu-profiling':
-        rsp.profile = agent().metrics.stopCpuProfiling();
-        rsp.notify = {
-          wid: worker.id,
-          pid: worker.pid,
-          cmd: 'cpu-profiling',
-          isRunning: false,
-        };
-        break;
+        agent().metrics.stopCpuProfiling(function(profile) {
```
Comment (rmg) [inline lib/targetctl.js]: The pattern used in the rest of this `switch` statement is to extract async responders in functions and do `return asyncResonder(req, res, callback);`.

File: lib/capabilities.js [hunk 43]
```diff
@@ -37,44 +36,17 @@ var capabilities = {
...
     name: 'Tracing',
     checks: [
       {
         check: function(cb) {
-          if (license('agent:tracing')) {
-            cb(true);
-          } else {
-            cb(false, 'Tracing license is missing or expired.');
-          }
+          cb(false, 'Appmetrics does not support tracing.');
```
Comment (rmg) [inline lib/capabilities.js]: Isn't this the tracing done in `lib/tracer.js` which you've updated?

File: lib/capabilities.js [hunk 44]
```diff
@@ -93,45 +65,21 @@ var capabilities = {
...
     checks: [
       {
         check: function(cb) {
-          var addon = require('strong-agent/lib/addon');
-          if (addon) {
-            cb(true);
-          } else {
-            cb(false, 'Strong-agent addon not compiled.');
-          }
+          cb(true);
```
Comment (rmg) [inline lib/capabilities.js]: I'm not sure if it would be clearer or not, but making the `checks` array empty would have the same result but with less overhead (not that it's measurable here).

File: lib/adapter.js (hunk 40 above)
Comment (sam-github) [inline lib/adapter.js]: pre-existing env var, so can't be changed

File: lib/capabilities.js (hunk 43 above)
Comment (sam-github) [inline lib/capabilities.js]: no :-(. this "tracing" is what we called the strongops "slow endpoints". The other tracing is the strong-trace/concurix tracing.
